sdk/evaluation/azure-ai-evaluation/azure/ai/evaluation/_evaluators/_retrieval/_retrieval.py

```diff
@@ -39,63 +40,71 @@ def __init__(self, model_config: dict):
         self._flow = AsyncPrompty.load(source=prompty_path, model=prompty_model_config)
 
-    async def __call__(self, *, conversation, **kwargs):
-        # Extract queries, responses and contexts from conversation
-        queries = []
-        responses = []
-        contexts = []
-
-        conversation = conversation.get("messages", None)
-
-        for each_turn in conversation:
-            role = each_turn["role"]
-            if role == "user":
-                queries.append(each_turn["content"])
-            elif role == "assistant":
-                responses.append(each_turn["content"])
-                if "context" in each_turn and "citations" in each_turn["context"]:
-                    citations = json.dumps(each_turn["context"]["citations"])
-                    contexts.append(citations)
-
-        # Evaluate each turn
-        per_turn_scores = []
-        history = []
-        for turn_num, query in enumerate(queries):
-            try:
-                query = query if turn_num < len(queries) else ""
-                answer = responses[turn_num] if turn_num < len(responses) else ""
-                context = contexts[turn_num] if turn_num < len(contexts) else ""
-
-                history.append({"user": query, "assistant": answer})
-
-                llm_output = await self._flow(
-                    query=query, history=history, documents=context, timeout=self._LLM_CALL_TIMEOUT, **kwargs
-                )
-                score = math.nan
-                if llm_output:
-                    parsed_score_response = re.findall(r"\d+", llm_output.split("# Result")[-1].strip())
-                    if len(parsed_score_response) > 0:
-                        score = float(parsed_score_response[0].replace("'", "").strip())
-
-                per_turn_scores.append(score)
-
-            except Exception as e:  # pylint: disable=broad-exception-caught
-                logger.warning(
-                    "Evaluator %s failed for turn %s with exception: %s", self.__class__.__name__, turn_num + 1, e
-                )
-
-                per_turn_scores.append(math.nan)
+    async def __call__(self, *, query, context, conversation, **kwargs):
+        if conversation:
+            # Extract queries, responses and contexts from conversation
+            queries = []
+            responses = []
+            contexts = []
+
+            conversation = conversation.get("messages", None)
+
+            for each_turn in conversation:
+                role = each_turn["role"]
+                if role == "user":
+                    queries.append(each_turn["content"])
+                elif role == "assistant":
+                    responses.append(each_turn["content"])
+                    if "context" in each_turn and "citations" in each_turn["context"]:
+                        citations = json.dumps(each_turn["context"]["citations"])
+                        contexts.append(citations)
```

Comment:
Is this logic still valid? I thought earch turn has a context field but it's string type, right?

Reply:
I think context is expected to be a string too. And no citation field.

Reply:
Here is my understanding of conversation what is happening conceptually: 

Single-turn mode: 
retrievel_eval(query: str, context: str)

Conversation mode: 
```
async def callback(
    messages: List[Dict], stream: bool = False, session_state: Optional[str] = None, context: Optional[Dict] = None
) -> dict:

    # global context: a dict with arbitary key-value pairs to store information outside of the per-turn messages (which store single-turn informaton like q/a pairs)   
    context: Optional[Dict] = {"global_key_A": "global_value_A", "global_key_B": "global_value_B"}

    messages_list = messages["messages"]
    # get last message
    latest_message = messages_list[-1]
    application_input = latest_message["content"]

    # "template_parameters" stores info outside of per-turn messages as arbitrary key-value pairs too 
    # for example, SIMULATOR outputs adversarial string to inject into your application input: 
    if "file_content" in messages["template_parameters"]:
        attack_string = messages["template_parameters"]["file_content"]
        # for jailbreak, application_input contains user query and/or third-party content
        # for rewrite/summarization
        application_input += attack_string

    aoai_response = example_application_response(application_input)
    message = {
        "content": aoai_response,
        "role": "assistant",
        # single-turn context: arbitrary key-value pairs storing info inside each turn
        # TODO: however, it is the EVALUATOR's responsibility to check this dict for the key-value pairs that fit the single-turn parameters! 
        "context": {"single_turn_key_A": "single_turn_value_A", "single_turn_key_B": "single_turn_value_B"}
    }

    messages["messages"].append(message)
    return {"messages": messages["messages"], "stream": stream, "session_state": session_state, "context": context}

```

Reply:
context is Dict[str, Any] in chat protocol, Neehar confirmed this

Reply:
I'll update the logic and add a test for both scenarios - when context is a string and when it's a dictionary.

Reply:
@diondrapeck, I also learned that the current behavior of all evaluators in conversation mode don’t look at other turns or global context. They only evaluate on single-turn parameters and output a list of score, plus an average score. You can validate with Billy. We should keep all evaluators touched in line with the current behavior.